### PR TITLE
Update SDKs to pull CLI from S3 URL.

### PR DESCRIPTION
**Pull request by sipsma**
Signed-off-by: Erik Sipsma <erik@sipsma.dev>

Brings the SDKs fully in line with the description of the new session behavior here: https://github.com/dagger/dagger/issues/3830#issuecomment-1332620438

Basically:
1. Prefer `DAGGER_SESSION_URL` if set, connect to it over http directly, no binary
2. Fallback to `_EXPERIMENTAL_DAGGER_CLI_BIN`, if set, as a pointer to the CLI, call the `session` subcommand on that
3. Finally, fallback to pulling a CLI binary from an URL encoded in the binary.
   * This is the same idea as when we used to pull the binary from the engine image w/ docker, except just an http request (with some extra complication due to tarball unpacking)

---

This PR is probably on the border of too large, but a lot of the changes are just re-arrangements of existing code, equivalent code in 3 SDKs and/or tests. If other prefer I can start by just merging the Go stuff and then send out separate PRs for python+nodejs.

---

Overall current state:
1. Go - solid afaict
5. Python - I did not put any effort into a large scale refactorization yet as I don't know the best way to approach it. I don't think we need the whole `@register_engine` thing anymore (I removed the equivalent of that in go+nodejs) but want to wait for your opinion first @helderco 
6. NodeJS - Ran into tons of headaches trying to get tar-related code working well (just fell back to making everything a file on the filesystem more or less, no in memory operations) and some other stuff related to test mocking. So it's ugly but works. Leaving comments on some questions I have.

Each of the SDKs have tests that setup an http server and use that to test the "pull from s3 url" code path. That's a good start but this code is still not tested against the real CLI deployed to S3 because we haven't done a release of the CLI yet that includes the new `session` subcommand that SDKs need.

So the last follow up (besides more cleanup of python+node) will be to either:
1. Setup the automation for pushing the CLI to s3 on every push to main
2. Do a release of the engine+cli, then test the SDKs against that, update as needed.

I'd greatly prefer 1 and talked to Gerhard about it a bit last week. It *seems* relatively straightforward, but I'll see how it goes when I try tomorrow.

### CODE REVIEW

**Comment by sipsma** [comment]
@helderco Updated the PR description with the backlink to the other issue and a summary, let me know what other questions come up!

**Comment by helderco** [comment]
@sipsma Thanks! 🙏

**Comment by helderco** [comment]
@sipsma We're we going to use `DAGGER_SESSION_PORT` instead? I need to catch up :)

File: sdk/nodejs/provisioning/bin.ts
Comment (slumbering) [reply sdk/nodejs/provisioning/bin.ts:340]: Looking at it now 👀

**Review by aluzzardi — APPROVED** [review]
Glad to see this! Definitely simplifies things a lot.

Overall LGTM, I'll let @helderco @slumbering @dolanor @TomChv dive deeper in the Node/Python parts.

I guess we could use this to "untangle" a little bit the dev environment at some point.

**Comment by helderco** [comment]
> @sipsma Weren't we going to use `DAGGER_SESSION_PORT` instead? I need to catch up 🙂

Ok, I see in https://github.com/dagger/dagger/issues/3830 (Implementation steps) that there's a task for it:

>  Switch `DAGGER_SESSION_URL` -> `DAGGER_SESSION_PORT`

File: sdk/nodejs/test/connect.spec.ts
Comment (slumbering) [inline sdk/nodejs/test/connect.spec.ts:62]: Not sure about the utility of those old* variables (from line 60-62). Same in the `after(()...`

Comment (slumbering) [reply sdk/nodejs/test/connect.spec.ts:62]: Oh I guess it's not reset process.env obj... 
I'm playing with Sinon.js that could stubs process.env values

Comment (slumbering) [reply sdk/nodejs/test/connect.spec.ts:62]: I'm still confused with `oldCLIHost` and `oldCLIScheme`

File: sdk/python/src/dagger/config.py
Comment (helderco) [reply sdk/python/src/dagger/config.py:21]: @sipsma I got a bit delayed. Should be ready by tomorrow but I'm not sure for today. I want to refactor the registration system but I can do it in a follow-up PR. Don't want to block you.

File: sdk/nodejs/test/connect.spec.ts
Comment (slumbering) [reply sdk/nodejs/test/connect.spec.ts]: @sipsma : I've just pushed a commit to remove them

File: sdk/python/src/dagger/engine/bin.py
Comment (helderco) [inline sdk/python/src/dagger/engine/bin.py:215]: Typo: 

```suggestion
    return f"{CLI_SCHEME}://{CLI_HOST}/dagger/releases/{cli_version}/checksums.txt"
```

**Review by helderco — APPROVED** [review]
As you say, can't really test default download logic but everything else seems to be working, just make sure to fix the typo.

I'm working on refactoring but I want to take the opportunity to fix a few more things so no need to wait for me if you need to move on, I can send a follow up PR.

File: sdk/python/src/dagger/_version.py
Comment (helderco) [inline sdk/python/src/dagger/_version.py:4]: Should update `.gitattributes` as well after this file rename.

File: sdk/nodejs/test/connect.spec.ts
Comment (sipsma) [reply sdk/nodejs/test/connect.spec.ts:62]: Hey sorry I didn't get back to this PR yesterday, but the reason for this is that CLI_HOST and CLI_SCHEME are set in the top level of the module, so changing them will affect any other tests. I think it technically doesn't hurt to leave them changed right now since no other test cases will hit the code paths that use them, but I'd rather leave them in since it could be very confusing in the future if someone is writing another test case and those vars have a different unexpected value. I went back to the original approach for now, but let me know what you think.

File: sdk/python/src/dagger/engine/bin.py
Comment (sipsma) [reply sdk/python/src/dagger/engine/bin.py]: Oops good catch, thanks, I missed this because I didn't realize we were skipping this test in mage (`-m not provision`), got rid of that now so the test runs and passes

File: sdk/nodejs/test/connect.spec.ts
Comment (slumbering) [reply sdk/nodejs/test/connect.spec.ts:62]: We should find another way in a near futur, like stubbing with `sinon.js` for instance.

**Review by slumbering — APPROVED** [review]

File: sdk/nodejs/test/connect.spec.ts
Comment (sipsma) [reply sdk/nodejs/test/connect.spec.ts:62]: Agreed, I tried a bunch of stuff like `rewire` but couldn't get it to work; I didn't think sinon would work because it would require we change our `Connect` method to accept a way of configuring all this, which means its configurable outside of tests too, which is what I was trying to avoid. Maybe it's just unavoidable though? Or I'm probably missing something obvious.

**Comment by sipsma** [comment]
For some reason the engine tests are stuck on:
```
Requested labels: ubuntu-22.04-16c-64g-600gb
Job defined at: dagger/dagger/.github/workflows/test.yml@refs/pull/4194/merge
Waiting for a runner to pick up this job...
```
Even after I tried restarting once. I'm gonna re-run those tests locally from a clean branch as a double check and will merge if they pass so we don't get blocked on that.

**Comment by sipsma** [comment]
> As you say, can't really test default download logic but everything else seems to be working, just make sure to fix the typo.
I'm working on refactoring but I want to take the opportunity to fix a few more things so no need to wait for me if you need to move on, I can send a follow up PR.

Thanks @helderco, that sounds great. The PR coming after this one will add publishes of the CLI on every push to main alongside a few updates to primarily the test case code in each SDK to support running against those CLI builds. As of right now the changes to the actual non-test code in SDKs is extremely small (just support for pulling from a different url path, similar to the way they can change host/scheme already), so shouldn't need to worry about conflicts.

**Comment by sipsma** [comment]
Engine tests passed locally, merging

**Pull request merged**
